File: AEDS2/estruturas_Abstratas_TAD/lista/listaCabeca.c
#include <stdio.h>
#include <stdlib.h>

typedef struct No {
    int dado;
    struct No *prox;
} No;

typedef struct {
    No *cabeca;
} Lista;

void inicializar(Lista *L) {
    L->cabeca = (No*) malloc(sizeof(No));
    L->cabeca->prox = NULL;
}

void inserirInicio(Lista *L, int valor) {
    No *novo = (No*) malloc(sizeof(No));
    novo->dado = valor;
    novo->prox = L->cabeca->prox;
    L->cabeca->prox = novo;
}

void remover(Lista *L, int valor) {
    No *ant = L->cabeca;
    No *atual = L->cabeca->prox;

    while (atual != NULL && atual->dado != valor) {
        ant = atual;
        atual = atual->prox;
    }

    if (atual != NULL) {
        ant->prox = atual->prox;
        free(atual);
    }
}

void mostrar(Lista *L) {
    No *aux = L->cabeca->prox;
    printf("Lista: ");
    while (aux != NULL) {
        printf("%d ", aux->dado);
        aux = aux->prox;
    }
    printf("\n");
}

int main() {
    Lista L;
    inicializar(&L);
    inserirInicio(&L, 10);
    inserirInicio(&L, 20);
    inserirInicio(&L, 30);
    mostrar(&L);
    remover(&L, 20);
    mostrar(&L);
    return 0;
}

File: AEDS2/estruturas_Abstratas_TAD/lista/listaCabeca.java
package AEDS.AEDS2.estruturas_Abstratas_TAD.lista;

class No {
    int dado;
    No prox;
    No(int d) { dado = d; prox = null; }
}

public class listaCabeca {
     No cabeca;

    listaCabeca() {
        cabeca = new No(0); // nó cabeça
    }

    void inserirInicio(int valor) {
        No novo = new No(valor);
        novo.prox = cabeca.prox;
        cabeca.prox = novo;
    }

    void remover(int valor) {
        No ant = cabeca;
        No atual = cabeca.prox;

        while (atual != null && atual.dado != valor) {
            ant = atual;
            atual = atual.prox;
        }

        if (atual != null) {
            ant.prox = atual.prox;
        }
    }

    void mostrar() {
        No aux = cabeca.prox;
        System.out.print("Lista: ");
        while (aux != null) {
            System.out.print(aux.dado + " ");
            aux = aux.prox;
        }
        System.out.println();
    }

    public static void main(String[] args) {
        listaCabeca l = new listaCabeca();
        l.inserirInicio(10);
        l.inserirInicio(20);
        l.inserirInicio(30);
        l.mostrar();
        l.remover(20);
        l.mostrar();
    }
}